### ISSUE TITLE
Questions/observations on requirements regarding to "links"

### ISSUE THREAD

**Issue by rouault** [opengeospatial/ogc-feat-geo-json#92]
- Recommendation 3 "/rec/types-schemas/type-links" , "The "links" array in any JSON-FG feature in the JSON document SHOULD include at least one link object where the "rel" member is set to "type"."  
    * so in a FeatureCollection of many Feature, each Feature would have this link object ? A bit of bloat if many features
    * the spec doesn't seem to explain what type of content is pointed by the link. I guess on purpose, but this doesn't look like something you can use for an automated process, but more for human consumption

- Recommendation 4 "/rec/types-schemas/schema-link-geojson"  .  The recmomandation says "If the JSON document is a GeoJSON feature" or "If the JSON document is a GeoJSON feature collection" whereas "Requirement 19 | /req/types-schemas/schema-link-jsonfg" mentions JSON-FG feature / feature-collections. I don't really get the point of the distinction. The standard addresses JSON-FG feature/feature-collections, and JSON-FG feature/feature-collections are GeoJSON one too, as being an extension.  While the phrasing is correct, the different language between Rec 4 and Req 19 is still a bit confusing. Either change the mention of GeoJSON to JSON-FG, or add a note saying the distinction here is mostly to be pedantic and that a JSON-FG feature[collection] is a GeoJSON one.

- We may get several links with rel=describedby and type=application/schema+json. From a consumer point of view, how do I know which one is the most "precise" ? Or am I (I = automater consumer perspective, like a GDAL driver) supposed to open each one and infer that from their content ?

- Recommendation 5  "/rec/types-schemas/schema-link-properties" doesn't really help if you have a document with several FeatureTypes. How do I know where to find the schema of a given FeatureType ? We'd ideally need something explicitly relating a value of FeatureType with the corresponding schema. As far as I know a single JSON schema cannot cleanly describe different feature types, contrary to a .xsd where in a single one you can have distinct <xs:element name="myFeatureType" type="my:myFeatureTypeType" substitutionGroup="gml:AbstractFeature"/> elements
In the building.json example (which has only one single FeatureType), I see that the hint is given in the "title"
```json
      {
         "href": "https://example.org/data/v1/collections/buildings/schemas/item?type=jsonfg",
         "rel": "describedby",
         "type": "application/schema+json",
         "title": "JSON Schema of building features"
      }
```
but that's not very helpful from an automated consumer perspective. Should there be a specific JSON object whose keys would be the featureType and the value the array of links that apply to it ?

<del>So overall, while .gml + .xsd is something that can be exploited "easily" (at least conceptually) from an automated reader point of view, JSON-FG links/schemas seem harder. (hum, not totally sure about what I've just written: I was forgetting for a second how painful making sense of .xsd can be :-))</del>

**Comment by cportele** [comment]
Thank you for the comments/questions, @rouault.

> Recommendation 3 "/rec/types-schemas/type-links" , "The "links" array in any JSON-FG feature in the JSON document SHOULD include at least one link object where the "rel" member is set to "type"."
> 
> * so in a FeatureCollection of many Feature, each Feature would have this link object ? A bit of bloat if many features

Good point. I guess framing this as a general recommendation is not a good idea. Using those link only makes sense, if there is a resource available that defines the feature type and provides additional information and if the supplier knows that users will benefit from this information. 

The main point of adding [text about links to a semantic type](https://docs.ogc.org/DRAFTS/21-045.html#_links_to_a_semantic_type) was to point out the existence of the IANA link relation type and before inventing something new the existing link relation should be used.

For homogenous feature collections, we might also just put a single link in the feature collection instead of the same link in each feature.

We should update the text to make this clearer.

> * the spec doesn't seem to explain what type of content is pointed by the link. I guess on purpose, but this doesn't look like something you can use for an automated process, but more for human consumption

Correct, at least in general. A community that maintains a feature concept or feature type register could be more specific. 

> Recommendation 4 "/rec/types-schemas/schema-link-geojson" . The recmomandation says "If the JSON document is a GeoJSON feature" or "If the JSON document is a GeoJSON feature collection" whereas "Requirement 19 | /req/types-schemas/schema-link-jsonfg" mentions JSON-FG feature / feature-collections. I don't really get the point of the distinction. The standard addresses JSON-FG feature/feature-collections, and JSON-FG feature/feature-collections are GeoJSON one too, as being an extension. While the phrasing is correct, the different language between Rec 4 and Req 19 is still a bit confusing. Either change the mention of GeoJSON to JSON-FG, or add a note saying the distinction here is mostly to be pedantic and that a JSON-FG feature[collection] is a GeoJSON one.

I see your point and I agree that we should change the reference in Recommendation 4 from "GeoJSON" to "JSON-FG".

> We may get several links with rel=describedby and type=application/schema+json. From a consumer point of view, how do I know which one is the most "precise" ? Or am I (I = automater consumer perspective, like a GDAL driver) supposed to open each one and infer that from their content ?

Technically, multiple such links imply an allOf wrapper around the referenced schemas. It depends on the information that the client is looking for, which is the most important one. I guess for a GDAL driver it would be information to identify the file as a JSON-FG document and information about the feature properties.

For the first aspect we have recently (not included in 0.1) introduced `conformsTo` ([link](https://docs.ogc.org/DRAFTS/21-045.html#metadata)) which should be better than trying to "sniff" information from a JSON Schema.

For the second aspect it would be the properties schema from Recommendation 5 ("/rec/types-schemas/schema-link-properties").

In OGC API Features we have recently made the decision to specify the schema of a collection with an encoding-agnostic JSON Schema that describes the logical model, not the GeoJSON or JSON-FG encoding. That schema then also describes the feature properties in other encodings like GML, CSV, FlatGeobuf, etc. If an API also wants to link to schemas for validation purposes, they can do that in accordance with the rules for the specific language (e.g., in the JSON Schema case, a link with `rel=describedby` and `type=application/schema+json`).

I think we should revisit the schema aspects in JSON-FG also with that in mind.

> Recommendation 5 "/rec/types-schemas/schema-link-properties" doesn't really help if you have a document with several FeatureTypes. How do I know where to find the schema of a given FeatureType ? We'd ideally need something explicitly relating a value of FeatureType with the corresponding schema. As far as I know a single JSON schema cannot cleanly describe different feature types, contrary to a .xsd where in a single one you can have distinct <xs:element name="myFeatureType" type="my:myFeatureTypeType" substitutionGroup="gml:AbstractFeature"/> elements
In the building.json example (which has only one single FeatureType), I see that the hint is given in the "title", [...] but that's not very helpful from an automated consumer perspective. Should there be a specific JSON object whose keys would be the featureType and the value the array of links that apply to it ?

Well, one approach to define a single schema for multiple feature types. A pattern could be:

```json
{
  "$schema": "https://json-schema.org/draft/2019-09/schema", 
  "oneOf": [
    {"$ref": "#/$defs/building" }, 
    {"$ref": "#/$defs/parcel" }, 
    {"$ref": "#/$defs/etc" }
  ], 
  "$defs": {
  ...
  }
}
```

This is still different from XML in the sense that XML has a element tag to identify the XSD definition to validate against and no such thing exists in JSON. To avoid ambiguities, we need to include the type name in the content - which is what the `featureType` member is about. This would be our own convention, and we need to add rules how to detect the schema fragment in the schema for a feature type, but I do not see another way, because JSON simply does not offer any standard mechanism.

One conclusion from all this is that we need to work more on the schema aspects to make it as easy and reliable as possible for clients like GDAL to extract schema information.

> hum, not totally sure about what I've just written: I was forgetting for a second how painful making sense of .xsd can be :-)

Indeed. But JSON Schema also can be quite complex and weird; and hard to parse. In OGC API Features we have a requirements and recommendations for a profile of JSON Schema that should be expressive enough, but relatively straightforward to parse (currently specified for queryables, but the idea is to apply this to schemas in general; [link](https://docs.ogc.org/DRAFTS/19-079r1.html#queryables), see `/req/queryables/get-queryables-response` and `/rec/queryables/queryables-schema`). In addition, the idea is to add additional keywords to declare properties as the feature id or the primary geometry property, etc.

As I said above, there is currently work going on in the API context and we need to make sure we approach this in a consistent way across the different standards, including JSON-FG.

**Comment by cportele** [comment]
Meeting 2023-08-21: The current thinking in the discussion was as described below. We will discuss again in the next meeting in two weeks.

- There are two reasons for using a schema: understanding the properties that a feature has / may have and validation of an instance. The first reason is considered to be the more important one.
- OGC API Features has decided to use a encoding-agnostic feature schema to describe the schema of a feature type - with additional information to tag special roles of properties, e.g., to identify the primary geometry, in order to be unambiguous how the properties are mapped to a feature encoding, e.g., GeoJSON or JSON-FG. That schema is represented as a JSON Schema, even though it will typically not be used for validation. The reason for using JSON Schema is the fact that JSON provides a straightforward way to express objects, arrays and values, has tool support and the fact that API developers are typically familiar with it (use in OpenAPI etc.). JSON-FG should consider to follow this approach, too, but need to think how we can declare the feature schema for the feature types in the document.
- With respect link(s) to JSON Schemas for validation (`rel='describedby'`) as well as for the link to a semantic type (`ref='type'`), we should remove normative language and simply point out that these standard link relation types can be used, if such links are relevant for the users of the data.
- The fact that a document is a JSON-FG document can be determined from the `conformsTo` array.
